9068 close account e2e

0xpatrickdev opened the pull request:
refs: #9068
refs: #9192

## Description
1. Includes e2e tests of different channel closing behaviors added in #9857:
 - ~~Automatically reopen a closed ICA channel (e.g. a packet timed out)~~ requires #9891 to complete. See  https://github.com/Agoric/agoric-sdk/pull/9864#issuecomment-2297232120
 - `CosmosOrchestrationAccount` (`IcaAccout`) holder can close their account
 - `CosmosOrchestrationAccount` (`IcaAccount`) holder can reopen their account

2. Includes `publish local and remote ibc addresses to vstorage` to facilitate querying the ICA account (channel) info from an off-chain client.
  - NOTE: if an account is a reopened, the localAddress and remoteAddress originally published to vstorage will be stale. Same with some of the values stored in `CosmosOrchestrationAccountKit`'s exo state. The ChainAddress will be the same, but the full address strings will contain new channelIDs. This is tech debt being taken on until #9066.
  
3. Includes test to ensure clients cannot successfully submit `MsgChanCloseInit` for ICA and Transfer channels

### Security Considerations
n/a

### Scaling Considerations
n/a

### Documentation Considerations
n/a

### Testing Considerations
This PR includes high fidelity tests with simulated chains.

### Upgrade Considerations
n/a

0xpatrickdev commented on multichain-testing/test/ica-channel-close.test.ts:
```diff
@@ -0,0 +1,345 @@
+
+    const { channels } = await retryUntilCondition(
+      () => remoteQueryClient.queryChannels(),
+      // @ts-expect-error ChannelSDKType.state is a string not a number
```
```suggestion
```

dckc commented:
taking myself off the reviewer list to focus on other things. LMK if that's not a good idea, @turadg and @LuqiPan

0xpatrickdev commented:
I removed the "auto reopen"  test scenario for now, as we need #9891 to complete it:

<details>
<summary>
unintentionalCloseAccountScenario (auto reopen)
</summary>

```ts
/** The channel is closed for an unexpected reason and should automatically reopen */
const unintentionalCloseAccountScenario = test.macro({
  title: (_, chainName: string) =>
    `Automatically reopen closed channel on ${chainName}`,
  exec: async (t, chainName: string) => {
    const config = chainConfig[chainName];
    if (!config) return t.fail(`Unknown chain: ${chainName}`);

    const {
      wallets,
      provisionSmartWallet,
      vstorageClient,
      retryUntilCondition,
      useChain,
      hermes,
    } = t.context;

    const agoricAddr = wallets[chainName];
    const wdUser1 = await provisionSmartWallet(agoricAddr, {
      BLD: 100n,
      IST: 100n,
    });
    t.log(`provisioning agoric smart wallet for ${agoricAddr}`);

    const doOffer = makeDoOffer(wdUser1);
    t.log(`${chainName} makeAccount offer`);
    const offerId = `${chainName}-makeAccount-${Date.now()}`;

    await doOffer({
      id: offerId,
      invitationSpec: {
        source: 'agoricContract',
        instancePath: [contractName],
        callPipe: [['makeOrchAccountInvitation']],
      },
      offerArgs: { chainName },
      proposal: {},
    });
    const currentWalletRecord = await retryUntilCondition(
      () => vstorageClient.queryData(`published.wallet.${agoricAddr}.current`),
      ({ offerToPublicSubscriberPaths }) =>
        Object.fromEntries(offerToPublicSubscriberPaths)[offerId],
      `${offerId} continuing invitation is in vstorage`,
    );
    const offerToPublicSubscriberMap = Object.fromEntries(
      currentWalletRecord.offerToPublicSubscriberPaths,
    );

    const accountStoragePath = offerToPublicSubscriberMap[offerId]?.account;
    t.assert(accountStoragePath, 'account storage path returned');
    const address = accountStoragePath.split('.').pop();
    t.log('Got address:', address);

    const {
      remoteAddress,
      localAddress,
    }: CosmosOrchestrationAccountStorageState =
      await vstorageClient.queryData(accountStoragePath);
    const { rPortID, rChannelID } = parseRemoteAddress(remoteAddress);
    const { lPortID, lChannelID, lConnectionID } =
      parseLocalAddress(localAddress);

    const dst = {
      chainId: chainInfo['agoric'].chainId,
      channelID: lChannelID,
      portID: lPortID,
      connectionID: lConnectionID,
    };
    const src = {
      chainId: useChain(chainName).chainInfo.chain.chain_id,
      channelID: rChannelID,
      portID: rPortID,
    };
    console.log(
      `Initiating channelCloseInit for dst: ${JSON.stringify(dst)} src: ${JSON.stringify(src)}`,
    );
    const closeChannelTx = hermes.channelCloseInit(chainName, dst, src);
    console.log('closeChannelExec', closeChannelTx);

    const remoteQueryClient = makeQueryClient(
      await useChain(chainName).getRestEndpoint(),
    );
    const { channel } = await retryUntilCondition(
      () => remoteQueryClient.queryChannel(rPortID, rChannelID),
      // @ts-expect-error ChannelSDKType.state is a string not a number
      ({ channel }) => channel?.state === 'STATE_CLOSED',
      'ICA channel closed from Hermes closeChannelInit',
      {
        retryIntervalMs: 300,
        maxRetries: 10,
      },
    );
    t.is(
      channel?.state,
      // @ts-expect-error ChannelSDKType.state is a string not a number
      'STATE_CLOSED',
      'closed state is observed',
    );

    const { channels } = await retryUntilCondition(
      () => remoteQueryClient.queryChannels(),
      // @ts-expect-error ChannelSDKType.state is a string not a number
      ({ channels }) => findNewChannel(channels, { rPortID, lPortID }),
      `ICA channel is reopened on ${chainName} Host`,
    );
    const newChannel = findNewChannel(channels, { rPortID, lPortID });
    t.log('New Channel after Reopen', newChannel);
    if (!newChannel) throw new Error('Channel not found');
    const newAddress = JSON.parse(newChannel.version).address;
    t.is(newAddress, address, `same chain address is returned - ${chainName}`);
    t.is(
      newChannel.state,
      // @ts-expect-error ChannelSDKType.state is a string not a number
      'STATE_OPEN',
      `channel is open on ${chainName} Host`,
    );
    t.not(newChannel.channel_id, rChannelID, 'remote channel id changed');
    t.not(
      newChannel.counterparty.channel_id,
      lChannelID,
      'local channel id changed',
    );
  },
});

test.serial(unintentionalCloseAccountScenario, 'cosmoshub');
test.serial(unintentionalCloseAccountScenario, 'osmosis');
```
</details>

This path is currently covered by unit tests with a `buildChannelCloseConfirmEvent` mock. The imperative `.deactivate()` and `.reactivate()` calls in this proposed E2E test also help firm up confidence around overall correctness.

turadg commented on packages/orchestration/src/exos/cosmos-orchestration-account.js:
```diff
@@ -178,23 +188,41 @@ export const prepareCosmosOrchestrationAccountKit = (
-
-      return { chainAddress, bondDenom, topicKit, ...rest };
+      // XXX consider parsing local/remoteAddr to portId, channelId, counterpartyPortId, counterpartyChannelId, connectionId, counterpartyConnectionId
+      // FIXME these values will not update if IcaAccount gets new values after reopening.
```
when? as part of some issue?

0xpatrickdev replied:
Updated the comment to mention https://github.com/Agoric/agoric-sdk/issues/9066
